### PR TITLE
Download and retrieve attachments

### LINKED ISSUE #1: A public user can download their customized, cloned skill list to a PDF

### LINKED ISSUE #2: A public user can download their customized, cloned skill list to an excel file
Related to #13. Any `OccupationStandard` can be downloaded in excel or PDF.

### PR DESCRIPTION
**Pull request by littleforest**
closes #1 
closes #2 

## Description
- Add endpoint to trigger pdf and excel generation
- Add occupation_standard#show endpoint to retrieve occupation_standard with pdf/excel info - including flag indicating whether pdf should be regenerated
- Add background job the generate excel file
- Remove requirement for occupation_standard_skill to belong to an occupation_standard_work_process

## QA
In admin dashboard
- [x] View occupation standard. Click "Generate Excel" button. Reload page. Should have link for csv file.
- [x] View csv file, make sure it looks correct.

In Postman
- [x] Request occupation standard 1 [per specs](https://rapid-skills-staging-pr-79.herokuapp.com/v1/docs/#retrieve-occupation-standard). 
- [x] Verify empty pdf and excel fields. 
- [x] Verify `should_generate_attachments` field is true
- [x] Trigger download for occupation standard 1 [per specs](https://rapid-skills-staging-pr-79.herokuapp.com/v1/docs/#pdf-excel-generation).
- [x] Request occupation standard 1 [per specs](https://rapid-skills-staging-pr-79.herokuapp.com/v1/docs/#retrieve-occupation-standard). 
- [x] Should see pdf and excel fields with download links
- [x] Verify `should_generate_attachments` field is false

### CODE REVIEW

**Comment by Pcushing** [comment]
@littleforest since the request to generate and request to get are separate, it seem like we need a status flag for the front-end. 

I think something like `generating_pdf` so the front-end knows if it should keep pinging for an updated version? 

It flips to `true` before firing the worker and inside the worker it flips to `false` just before completion?

**Comment by littleforest** [comment]
@Pcushing @jhash what about returning a `pdf_last_updated_at` type of field, and then the client could do a comparison to see if the file has been updated since the request was made?  For the flag idea, if multiple people are requesting the file, then the flag would not be very accurate.

**Comment by Pcushing** [comment]
> For the flag idea, if multiple people are requesting the file, then the flag would not be very accurate.

Good point.

> what about returning a `pdf_last_updated_at` type of field, and then the client could do a comparison to see if the file has been updated since the request was made

I like this. I think we can take this a step further to see if we need to regenerate at all rather than doing it every time. If `pdf_last_updated_at` isn't the same as `updated_at`, it seems we need to regenerate. Otherwise, we shouldn't have to create a new PDF.

When viewing an `OccupationStandard` that someone else's created, presumably, these aren't changing much so we should just be able to get the `pdf_file_url` most of the time. For an `OccupationStandard` a user is working on, it's more likely we'll need to generate in the midst of the workflow.

**Comment by littleforest** [comment]
@Pcushing this is ready for re-review.  I added the `pdf_created_at` timestamp to the occupation standard response, but also a `should_generate_pdf` boolean that the client can use to trigger the download.  In the PDF generation background job, I set it to just exit if the pdf does not need to be updated.

**Comment by littleforest** [comment]
@Pcushing this is ready for re-review, changes include both pdf and excel updates.

File: app/models/occupation_standard.rb [hunk 1]
```diff
@@ -49,6 +51,36 @@ def clone_as_unregistered!(creator_id:, organization_id:)
...
 
+  def should_generate_attachment?(kind)
+    # Occupation standard updated_at timestamp gets set milliseconds after
+    # attachment created_at timestamp.
+    !send(kind).attached? || send(kind).created_at < updated_at - 1.second
+  end
+
+  def to_csv
+    CSV.generate do |csv|
+      csv << %w(rapids_code organization_title occupation_standard_title type work_process_title work_process_description work_process_hours work_process_sort skill skill_sort)
```
Comment (Pcushing) [inline app/models/occupation_standard.rb]: Two things:

1. Let's include the `onet_code` here after `rapids_code`.

2. Just a style comment on this, but can we pull these headers out into a constant like 

```CSV_HEADERS = %w(rapids_code organization_title occupation_standard_title type work_process_title work_process_description work_process_hours work_process_sort skill skill_sort).freeze``` 

at the top of the model file?  Then we just have `csv << CSV_HEADERS` in the `#to_csv` method?

File: app/models/occupation_standard.rb [hunk 2]
```diff
@@ -49,6 +51,36 @@ def clone_as_unregistered!(creator_id:, organization_id:)
...
+      occupation_standard_work_processes.includes(:work_process, :occupation_standard_skills, :skills).each do |oswp|
```
Comment (Pcushing) [inline app/models/occupation_standard.rb]: If you think we'll end up using this `includes` in a few places (per discussion of how the api should work on the call today), should we make this a scope so we dont have to write the whole thing out each time?

File: app/models/occupation_standard.rb [hunk 3]
```diff
@@ -49,6 +51,36 @@ def clone_as_unregistered!(creator_id:, organization_id:)
...
+        array = [rapids_code, organization_title, title, type.gsub('Standard', ''), oswp.work_process_title, oswp.work_process_description, oswp.hours, oswp.sort_order]
```
Comment (Pcushing) [inline app/models/occupation_standard.rb]: Can we isolate this into its own method, too? Potentially handling the check for `oswp.skills` in the same place?

```
def csv_row(owsp)
 [rapids_code, organization_title, title, type.gsub('Standard', ''), oswp.work_process_title, oswp.work_process_description, oswp.hours, oswp.sort_order]
end
```

Main goal with these comments is to make this `#to_csv` easier to read. Viewing on github compounds readability issues by squeezing into more rows in the diff, but this also sets us up to more easily support a `to_xls` option later.

File: app/models/occupation_standard.rb [hunk 4]
```diff
@@ -49,6 +51,36 @@ def clone_as_unregistered!(creator_id:, organization_id:)
...
+        if oswp.skills.any?
+          oswp.occupation_standard_skills.each do |oss|
+            csv << array + [oss.skill_description, oss.sort_order]
+          end
+        else
+          csv << array
+        end
+      end
+
+      occupation_standard_skills.where(occupation_standard_work_process: nil).each do |oss|
```
Comment (Pcushing) [inline app/models/occupation_standard.rb]: Given this row, maybe it's more like 

`#work_process_row` and `#skill_row`

**Comment by Pcushing** [comment]
@littleforest looks good. I'll take care of the QA steps after we tidy up the `to_csv` comments. All of https://github.com/WorkHands/WorkHands csv code looks exactly like this, but I contracted on a project that did a better job isolating this so they could more easily deliver csv and excel. 

Let's see if we can do that from the start with rapidskills.

**Comment by littleforest** [comment]
@Pcushing thanks for the great suggestions.  This is ready for re-review.

**Comment by Pcushing** [comment]
@littleforest good-to-go

**Pull request merged**